File: GoLstar/lstarPack/lstar.go
package lstarPack

import (
	"fmt"
)

// lstar struct
type lstar struct{
	alphabet [2]string
	mq mem_query
	eq equiv_query
	num_mq int
	num_eq int
}

// query type
type mem_query func(string) bool

type equiv_query func(Dfa) string

// create a new lstar object
func NewLstar(alpha [2]string, in_mq mem_query, in_eq equiv_query) lstar{
	return lstar{alphabet: alpha, mq: in_mq, eq: in_eq}
}

// learn function from DFA
func (l lstar) Learn_dfa() Dfa{
	l.num_eq = 0
	l.num_mq = 0
	Q := []string{""}
	T := []string{""}
	tf := map[string]map[string]string{}
	iter := 1

	for {
		Q, tf = l.close(Q, T)

		final_states := []string{}
		for _, state := range Q {
			l.num_mq = l.num_mq + 1
			if l.mq(state) == true {
				final_states = append(final_states, state)
			}
		}

		// create hyopthesis dfa
		h_DFA := Dfa{States: Q, Alphabet: l.alphabet, Start: "", Tf: tf, Accepting: final_states}

		l.num_eq = l.num_eq + 1
		ret := l.eq(h_DFA)

		if ret == "None"{
			fmt.Println("Number of Membership Queries:", l.num_mq)
			fmt.Println("Number of Equivalence Queries:", l.num_eq)
			return h_DFA
		}

		new_state, new_test_word := l.add_test_word(tf, ret)

		if slice_contains(Q, new_state) == false{
			Q = append(Q, new_state)
		}

		if slice_contains(T, new_test_word) == false{
			T = append(T, new_test_word)
		}

		iter++
	}
}

// check if two strings are indistinguishable given a set
func (l lstar) are_indistinguishable(t_set []string, u string, v string) bool{
	for _, t := range t_set{
		l.num_mq = l.num_mq + 2
		if l.mq(u + t) != l.mq(v + t){
			return false
		}
	}
	return true
}

// close the two provided sets of strings
func (l lstar) close(q_set []string, t_set []string) ([]string, map[string]map[string] string){
	tf := map[string]map[string]string{}
    q_copy := make([]string, len(q_set))
	copy(q_copy, q_set)
	
	for i := 0; i < len(q_set); i++ {

		for _, a := range l.alphabet {
			
			q := q_set[i]

			for _, r := range q_set {

				if l.are_indistinguishable(t_set, (q + a), r) == true{

					if nest_map_contains(tf, q) == false {
						tf[q] = map[string]string{}
					} 
					
					tf[q][a] = r
					break
				}	
			}

			if nest_map_contains(tf, q) == false {
				tf[q] = map[string]string{}
			}

			if map_contains(tf[q], a) == false{
				q_copy = append(q_copy, q+a)
				tf[q][a] = q+a
			}
		}
	}

	return q_copy, tf
}

// create a new test word from a given counter-exampling
func (l lstar) add_test_word(tf map[string]map[string] string, ce string) (string, string){
	q := ""
	i := 0

	for {
		ele, exists := tf[q][string(ce[i])]
		l.num_mq = l.num_mq + 2

		if exists {
			
			if l.mq(ele + ce[i+1:]) != l.mq(ce){
				return (q + string(ce[i])), ce[i+1:]				
			}

		} else {
			return (q + string(ce[i])), ce[i+1:]
		}

		q = ele
		i++
	}
}

File: InterfaceLstar/PyFolder/lstar.py
# custom made DFA class 
from .dfa import DFA
import socket

# will create an lstar object
def init_lstar(ip, port, buffer_size):
    lstar_obj = Lstar()

    ip_str = ip.decode()
    port_str = port.decode()

    lstar_obj.init_s(ip_str, port_str, buffer_size)

    return lstar_obj
    
# begins the learning process given an lstar object
def learn(lstarObj, alphabet = ["0", "1"]):
    json_DFA = lstarObj.learn_DFA(alphabet)
    
    return bytes(json_DFA, "utf-8")

# lstar class
class Lstar:
    # initialise an object and establish a connection to a server
    def init_s(self, ip, port, buffer_size):
        self.ip = ip
        self.port = port
        self.buffer_size = int(buffer_size)
    
        self.s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.s.connect((ip, int(port)))

    def ask_mq(self, word):
        self.num_mq += 1

        query = "mq " + word
        self.s.sendall(query.encode())
        data = self.s.recv(self.buffer_size).decode()

        if data == "true":
            return True
        else:
            return False 

    # ask equivalence query
    def ask_eq(self, json_dfa):
        self.num_eq += 1
        query = "eq " + json_dfa
        
        self.s.sendall(query.encode())
        data = self.s.recv(self.buffer_size)

        dataStr = data.decode()
        if dataStr == "empty":
            dataStr = ""

        return dataStr

    # send end to the tcp server to close connection
    def end_ws(self):
        self.s.send("end".encode())

    # will learn a DFA based on the examples and regular expression provided
    def learn_DFA(self, alphabet = ['0', '1']):
        self.num_mq = 0
        self.num_eq = 0

        self.alphabet = alphabet
        self.iter = 0
        
        qQ = [""]
        tT = [""]

        while True:
            # expand states and transition function 
            qQ, tf = self.close(qQ, tT)
            # print("\tAfter Close")

            # get final states in q
            final_states = []
            for pos in qQ:
                if self.ask_mq(pos):
                    final_states.append(pos)

            # create hypothesis dfa
            self.hDFA = DFA(qQ, self.alphabet, "", tf, final_states)

            # determine if dfa properly classifies the examples provided
            res = self.ask_eq(self.hDFA.export_JSON())

            # if so return the hypothesis dfa
            if res == "None": # if res True then hDFA classifies L (or the examples provided)
                self.end_ws()
                print("Number of Membership Queries:", self.num_mq)
                print("Number of Equivalence Queries:", self.num_eq)
                return self.hDFA.export_JSON()
            
            # consider the counter example and create a new state and test word to be considered
            new_state, new_test_word = self.add_test_word(qQ, tf, res) # res should contain a counterexample that H does not classify
            
            if new_state not in qQ:
                qQ.append(new_state)

            if new_test_word not in tT:
                tT.append(new_test_word)

            self.iter += 1


    # function to check for t equivalence
    def are_indistinguishable(self, T, u, v):
        # go through the suffixes stored in T
        for t in T:
            if self.ask_mq( (u+t) ) != self.ask_mq( (v+t)):
                return False

        return True

    # main loop sequence in l star
    def close(self, qQ, tT):
        # create a new empty tf, i to go through all states, create a copy of states to avoid infinite loop
        tf = {}
        i = 0
        newQ = qQ.copy()

        # go through all states
        while i < len(qQ):
            
            # go through the alphabet
            for a in self.alphabet:
                
                # get a new state
                q = qQ[i]

                # go through all states again
                for r in qQ:

                    # check if a state and a character in the alphabet lead to another state meaning that these two states are T-equivalent
                    if self.are_indistinguishable(tT, (q + a), r):
                        
                        # avoid key error if this is the first instance of q in tf
                        if q not in tf:
                            tf[q] = {}

                        # link the two states in the tf
                        tf[q][a] = r    
                        break
                
                # avoid key error if first instance of q in tf
                if q not in tf:
                    tf[q] = {}

                # if a not in the tf[q] append the state and letter as a new state and create a new entry in tf
                if a not in tf[q]:
                    newQ.append(q + a)
                    tf[q][a] = q + a

            i += 1
        return newQ, tf

    # used to treat a new counter example correctly
    # will go through the transition function accordingly depending on the counter example provided
    def add_test_word(self, Q, tf, ce):
        # q will store the string as we go through it and i will store the index
        q = ""
        i = 0

        # infinite loop
        while True:
            # try except that will trigger if the state is not yet in tf, same thing is carried out if triggered
            # check if the transition from the current state and the rest of string leads to the counter example
            try:
                if self.ask_mq( (tf[q][ce[i]] + ce[i+1:]) ) != self.ask_mq(ce):
                    return (q + ce[i]), ce[i+1:]

            except KeyError:
                return (q + ce[i]), ce[i+1:]                

            # if not store q as the current part of the string and update the i
            q = tf[q][ce[i]]
            i += 1